src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs

```diff
@@ -20,8 +20,64 @@ internal enum NullableAnnotation : byte
         Nullable,     // Explicitly set by flow analysis
     }
 
+    /// <summary>
+    /// The nullable state of an rvalue computed in <see cref="NullableWalker"/>.
+    /// When in doubt we conservatively use <see cref="NullableFlowState.NotNull"/>
+    /// to minimize diagnostics.
+    /// </summary>
+    internal enum NullableFlowState : byte
+    {
+        NotNull,
+        MaybeNull
+    }
+
+    /// <summary>
+    /// A type and its corresponding flow state resulting from evaluating an rvalue expression.
+    /// </summary>
+    [DebuggerDisplay("{GetDebuggerDisplay(), nq}")]
+    internal struct TypeWithState
+    {
+        public TypeSymbol Type { get; }
+        public NullableFlowState State { get; }
+        public bool HasNullType => Type is null;
+        public bool MaybeNull => State == NullableFlowState.MaybeNull;
+        public bool NotNull => State == NullableFlowState.NotNull;
+        public bool IsDefault => Type is null;
```

Comment:
>IsDefault [](start = 20, length = 9)

It looks like if `Type` is null, then we consider the whole structure "default" and thus we shouldn't look at the `State`. Is that correct?
If so, can this be encoded with assertions?